src/blog/0.3/linear-types-borrowing.vmd

```diff
@@ -0,0 +1,673 @@
+
+
+Generational references seem to occupy a sweet spot because they allow objects to be linear (they're owned by only one place), yet allow shared mutability in a way that doesn't artificially extend the lifetime of the object.
+
+
+Since we're combining linear types with aliasing, we'll call this the *linear-aliasing model*.
+
+
+The mark of a good language is when its features compose well and don't conflict with each other. It seems we're on the right track with this model, as it composes linear types with shared mutability pretty well.
+
+
+<slice/>
+
+# Matches Made in Heaven
+
+A pleasant surprise, we're back to where we started!
+
+
+But we arrive with a new understanding:
+
+ * We can code in a "linear-first" fashion, and then only add references if we want some extra flexibility.
+ * Alternatively, we can code with a lot of references, and then bring them down to *zero* where we want some more speed.
+
+In other words, we now have complete control over the flexibility vs. performance tradeoff, something I've never seen a language do well.
+
+
+It was quite a shock, realizing that Vale could do this the entire time.
+
+
+And it made me wonder, what does this mean for [regions](/blog/zero-cost-borrowing-regions-overview), Vale's other mechanism for eliminating generation checks?
+
+
+Then I realized, they fit _perfectly_ together. We can use regions to look at a function's inputs in a zero-cost way, and use this "linear style" for any new data the function creates.
+
+I tried this out in a little [roguelike game](https://github.com/Verdagon/RegionsBenchmarks/blob/main/tiny-roguelike/roguelike.vale). The vast majority of generation checks were eliminated by regions, and the rest were eliminated by this new linear style.
+
+[# E: Should I go into more detail about this balance, or save it for a future article?]
+
+
+You can try this too! Just invoke the Vale compiler with the `--print_mem_overhead true` flag to make the program print out how many generation checks you can still eliminate (and if you want to use regions too, check out the experimental branch [described here](/blog/first-regions-prototype)).
+
+
+
+<slice/>
+
+# Conclusion and Next Steps
+
+For those of us that like to squeeze every little last drop of performance out of our programs, this is a pretty amazing realization.
+
+
+But ironically, that's a small minority. For the vast majority of Vale programs, generational references aren't a noticeable source of overhead, and none of this will matter. And that's a good thing; one of Vale's main strengths is that you can just make your program, and things will just work and will be fast.
+
+
+However, this does open up Vale to some new domains. Since we can program in a way that doesnt require generation checks, that means that Vale could be used in new niches, like embedded systems, operating system kernels, and safety-critical software. That's pretty exciting.
```

Comment:
Feels like you need one final sentence of hype, like "this is an awesome discovery and we're can't wait to think of what else it could lead to" or something